apps/prairielearn/src/sprocs/users_select_or_insert.sql

```diff
@@ -76,7 +80,8 @@ BEGIN
             AND ap.name = authn_provider_name;
 
         IF NOT FOUND THEN
-            RAISE EXCEPTION '"%" authentication provider is not allowed for institution "%"', authn_provider_name, institution.long_name;
+            result := 'invalid_authn_provider';
```

Comment:
I wonder if it may be better to use `RAISE EXCEPTION ... WITH ERRCODE` (or something similar, like `DETAIL` or `HINT`) similar to what is done with the status code, then catch the exception and get the info from `err.data.sqlError`. I think this  kind of construct is a bit weird...

Reply:
That is in fact how I originally implemented things, see https://github.com/PrairieLearn/PrairieLearn/pull/6151/commits/22a4c0ec6a26f30ee48eb9857891ec8e79ae9d26. I changed the implementation as discussed here: https://github.com/PrairieLearn/PrairieLearn/pull/6151#discussion_r950714886

One change that occurred since this PR was introduced is that we now only really have a single site where `users_select_or_insert` is called, which makes the "catch error and parse the institution ID out of it" approach a little more tenable. I'm inclined to not make a change here, though if @mwest1066 agrees with you, I'll reconsider.

Reply:
The difference between my suggestion and the code you originally had was that your code was based on the error message, while I'm suggesting using another field (e.g., code, detail, hint) to identify the error type. This should be more robust against changes in wording of the exception text. But it's up to you.

Reply:
I don't see a meaningful difference between those and the exception message; we'd still have to parse an institution ID out of *something*. I suppose we could directly set `detail` to an institution ID, but that almost feels worse than an explicit `user_instititution_id` in the results.

Reply:
To be clear, I'm ok with institution ID as an output parameter, it's just the `result` column that I have a problem with.

Reply:
Unless I'm missing something, a sproc cannot both raise an exception and return a value. We only actually need the institution ID when we're handling a failure case (that is, when you're proposing we raise an exception).

Reply:
Ah, right, didn't think of that.